RFC/src/RFC-0241_AtomicSwapXMR.md

```diff
@@ -0,0 +1,789 @@
+can claim it with the lapse transaction. The counterparty can then claim the Monero; however, this will open up an 
+attack vector to enable either party to claim the Monero while claiming the Tari. Either party could trivially pull off 
+such a scheme by performing a front-running attack and having a bit of luck. The counterparty monitors all broadcast 
+transactions to base nodes. Upon identifying the lapse transaction, they do two things; in quick succession, broadcast 
+their lapse transaction and the transaction to claim the Monero, both with sufficiently high fees. Base nodes will 
+prefer to mine transactions with the higher fees, and thus the counterparty can walk away with both the Tari and the 
+Monero.
+
+## Key construction
+
+Using [multi-signatures](https://reyify.com/blog/flipping-the-scriptless-script-on-schnorr) with Schnorr signatures, we
+need to ensure that the keys are constructed so that key cancellation attacks are not possible. To do this, we create new
+keys from the chosen public keys \\(K_a'\\) and \\(K_b'\\)
+
+$$
+\begin{aligned}
+K_a &=  \hash{\hash{K_a' \cat K_b'} \cat K_a' } * K_a' \\\\
+k_a &=  \hash{\hash{K_a' \cat K_b'} \cat K_a' } * k_a' \\\\
+K_b &=  \hash{\hash{K_a' \cat K_b'} \cat K_b' } * K_b' \\\\
+k_b &=  \hash{\hash{K_a' \cat K_b'} \cat K_b' } * k_b' \\\\
+\end{aligned}
+\tag{1}
+$$
+
+## Method 1
+
+### Detail
+
+This method relies purely on TariScript to enforce the exposure of the private Monero aggregate keys. The script forces 
+the spending party to supply their Monero private key part as input data to the script, evaluated via the operation 
+`Ed25519Point`. The simplicity of this method lies therein that the spending party creates all transactions on their 
+own. Bob requires a pre-image from Alice to complete the swap transaction; Alice needs to verify that Bob published the 
+Monero transaction and that everything is complete as they have agreed. If she is happy, she will provide Bob with the 
+pre-image to immediately claim the Tari UTXO.
+
+### TariScript
+
+The Script used for the Tari UTXO is as follows:
+``` TariScript,ignore
+   Ed25519Point
+   CheckHeight(height_1)
+   LtZero
+   IFTHEN
+      PushPubkey(Xm_b)
+      EqualVerify
+      HashSha256 
+      PushHash(HASH256{pre_image})
+      EqualVerify
+      PushPubkey(K_{Sb})
+   Else
+      CheckHeight(height_2)
+      LtZero
+      IFTHEN
+         PushPubkey(Xm_a)
+         EqualVerify
+         PushPubkey(K_{Sa})
+      Else
+         PushPubkey(Xm_b)
+         EqualVerify
+         PushPubkey(K_{Sb})
+      ENDIF
+   ENDIF
+```
+
+Before `height_1`, Bob can claim the Tari UTXO by supplying `pre_image` and his private Monero key part `x_b`. After 
+`height_1` but before `height_2`, Alice can claim the Tari UTXO by supplying her private Monero key part `x_a`. After 
+`height_2`, Bob can claim the Tari UTXO by supplying his private Monero key part `x_b`.
+
```

Comment:
I think we need a short discussion about choosing safe values for `height_1` and `height_2`.

Reply:
That's an implementation detail I don't think we need in here. But I am open to what the values can be. I have not personally put any thought in the actual heights apart from height_1 should be bigger than height_2.

In the XTR-BTC swap demo, I used a day for the height.  But there was no actual reasoning behind choosing that specific value.

Reply:
Maybe something like:

_The first lock height `height_1` should give ample time for (a) Alice to post the Tari UTXO transaction and for it to be mined with a safe number of confirmations, (b) for Bob to post the Monero transaction and for it to be mined with a safe number of confirmations, (c) for Alice to give the pre-image to Bob and for him to post the swap transaction and for it to be mined with a safe number of confirmations. The second lock height `height_2` should give ample time for Alice after  `height_1` to re-claim her Tari._

_As a guideline, a "quick" atomic swap could be 6 hours for `height_1` and 12 hours for `height_2`, whereas a "conservative" time frame could be 24 hours and 48 hours respectively._